xla/service/gpu/gpu_norm_runner.cc

```diff
@@ -26,38 +26,60 @@ namespace xla {
 namespace gpu {
 
 Status RunGpuNorm(const gpu::GpuNormConfig& config,
-                  const se::DeviceMemoryBase& input_buffer,
+                  const se::DeviceMemoryBase& x_buffer,
                   const se::DeviceMemoryBase& scale_buffer,
-                  const se::DeviceMemoryBase& bias_buffer,
-                  const se::DeviceMemoryBase& output_buffer,
+                  const se::DeviceMemoryBase& y_or_dx_buffer,
+                  std::optional<se::DeviceMemoryBase> bias_buffer,
+                  std::optional<se::DeviceMemoryBase> dy_buffer,
                   std::optional<se::DeviceMemoryBase> expectation_buffer,
                   std::optional<se::DeviceMemoryBase> norm_factor_buffer,
+                  std::optional<se::DeviceMemoryBase> dscale_buffer,
+                  std::optional<se::DeviceMemoryBase> dbias_buffer,
                   const se::DeviceMemoryBase& scratch_memory,
                   se::Stream* stream, RunNormOptions options) {
   se::dnn::LazyOpRunner<se::dnn::NormOp>* lazy_runner =
       options.norm_runner->AsNormRunner();
   std::optional<se::dnn::LazyOpRunner<se::dnn::NormOp>> local_runner;
 
-  se::dnn::NormOp::Config ln_config{config.epsilon,
-                                    config.input_descriptor,
+  TF_ASSIGN_OR_RETURN(se::dnn::NormKind kind,
+                      GetDNNNormKindFromCudnnNormKind(config.kind));
+
+  se::dnn::NormOp::Config ln_config{kind,
+                                    config.epsilon,
+                                    config.x_descriptor,
                                     config.scale_descriptor,
+                                    config.y_or_dx_descriptor,
                                     config.bias_descriptor,
-                                    config.output_descriptor,
+                                    config.dy_descriptor,
                                     config.expectation_descriptor,
-                                    config.norm_factor_descriptor};
+                                    config.norm_factor_descriptor,
+                                    config.dscale_descriptor,
+                                    config.dbias_descriptor};
   TF_ASSIGN_OR_RETURN(auto* runner,
                       lazy_runner->GetOrCreateRunner(ln_config, stream));
 
   std::vector<se::DeviceMemoryBase> operands;
-  operands.emplace_back(input_buffer);
+  operands.emplace_back(x_buffer);
   operands.emplace_back(scale_buffer);
-  operands.emplace_back(bias_buffer);
-  operands.emplace_back(output_buffer);
-  if (expectation_buffer) {
+  operands.emplace_back(y_or_dx_buffer);
+
+  // The optional operands are composed of inputs followed by outputs of the
```

Comment:
The word "optional" initially made me think some operands are not required. Call this std::optional